clippy_lints/src/needless_borrow.rs

```diff
@@ -85,50 +138,153 @@ impl<'tcx> LateLintPass<'tcx> for NeedlessBorrow {
+                then {
+                    let map = cx.tcx.hir();
+                    let mut iter = map.parent_iter(pat.hir_id);
+                    let expr_ctxt = loop {
+                        // Find the context shared by the pattern, and it's scope use site.
+                        match iter.next() {
+                            None => {
+                                // This shouldn't happen, but mark this as unfixable if it does.
+                                self.ref_locals.insert(id, None);
+                                return;
+                            }
+                            Some((_, Node::ImplItem(ImplItem { span, .. })
+                                | Node::TraitItem(TraitItem { span, .. })
+                                | Node::Item(Item { span, .. })
+                                | Node::Block( Block { span, ..})
+                                | Node::Expr( Expr { span, .. }))) => break span.ctxt(),
+                            _ => (),
+                        }
+                    };
+
+                    if pat.span.ctxt() == expr_ctxt {
+                        let mut app = Applicability::MachineApplicable;
+                        let snip = snippet_with_context(cx, name.span, expr_ctxt, "..", &mut app).0;
+                        self.current_body = self.current_body.or(cx.enclosing_body);
+                        self.ref_locals.insert(
+                            id,
+                            Some(RefPat {
+                                lint: NEEDLESS_BORROW,
+                                spans: vec![pat.span],
+                                ctxt: expr_ctxt,
+                                app,
+                                replacements: vec![(pat.span, snip.into())],
+                            }),
+                        );
+                    } else {
+                        // The context of the pattern is different than the context using the binding.
+                        // Changing the pattern might affect other code which needs the ref binding.
+                        // Mark the local as having been seen, but not fixable.
+                        self.ref_locals.insert(id, None);
+                    }
+                }
+            }
         }
-        if_chain! {
-            if let PatKind::Binding(BindingAnnotation::Ref, .., name, _) = pat.kind;
-            if let ty::Ref(_, tam, mutbl) = *cx.typeck_results().pat_ty(pat).kind();
-            if mutbl == Mutability::Not;
-            if let ty::Ref(_, _, mutbl) = *tam.kind();
-            // only lint immutable refs, because borrowed `&mut T` cannot be moved out
-            if mutbl == Mutability::Not;
-            then {
+    }
+
+    fn check_body_post(&mut self, cx: &LateContext<'tcx>, body: &'tcx Body<'_>) {
+        if Some(body.id()) == self.current_body {
+            for pat in self.ref_locals.drain(..).filter_map(|(_, x)| x) {
```

Comment:
`check_body_post` is called after every closure and after every nested function. This will cause you to flush your data before seeing more binding usages. You probably need a stack to handle nested bodies.

```rust
let ref b = 1;
fn f() { todo!() } // check_body(_post) called here
let _ = || todo!(); // check_body(_post) called here
println!("{}", b);
```

Reply:
The body of the first binding seen is recorded so nothing is done until that body is exited. At that point none of the bindings seen to that point can be used any further.

Reply:
Ohh I see. Neat trick!